src/controls/Widget.ts

```diff
@@ -107,6 +108,54 @@ export class Widget_markdown implements IRequest<'markdown', Widget_markdown_opt
     get result(): Widget_markdown_output { return this.state }
 }
 
+// 🅿️ custom ==============================================================================
+export type Widget_custom_componentProps_ui = {
+    image: (props: {     
+        size?: string | undefined;
+        img: MediaImageID; 
+    }) => JSX.Element
+}
+export type Widget_custom_componentProps<T = unknown> = {
+    value: undefined | T
+    onChange: (value: undefined | T) => void
+    ui: Widget_custom_componentProps_ui
+}
+
+export type Widget_custom_opts<TViewState>  = ReqInput<{ customComponent: (props: Widget_custom_componentProps<TViewState>) => JSX.Element }>
+export type Widget_custom_serial<TViewState> = StateFields<{ type: 'custom', active: true; componentViewState: undefined | TViewState }>
+export type Widget_custom_state<TViewState>  = StateFields<{ type: 'custom', active: true; componentViewState: undefined | TViewState }>
+export type Widget_custom_output<TViewState> = undefined | TViewState
+export interface Widget_custom<TViewState> extends IWidget<'custom', Widget_custom_opts<TViewState>, Widget_custom_serial<TViewState>, Widget_custom_state<TViewState>, Widget_custom_output> {}
+export class Widget_custom<TViewState> implements IRequest<'custom', Widget_custom_opts<TViewState>, Widget_custom_serial<TViewState>, Widget_custom_state<TViewState>, Widget_custom_output> {
+    isOptional = false
+    id: string
+    type: 'custom' = 'custom'
+    state: Widget_custom_state<TViewState>
+
+    get customComponent(): ((props: Widget_custom_componentProps<TViewState>) => JSX.Element) {
+        return this.input.customComponent;
```

Comment:
[I can do that after merging]: we should wrap the component in observer automatically